### PR TITLE
8289094: [JVMCI] reduce JNI overhead and other VM rounds trips in JVMCI

### PR DESCRIPTION
**Pull request by dougxc**
The interface between HotSpot and libjvmci is implemented via JNI. Every time HotSpot C++ code needs to access libjvmci objects, it requires 4 VM transitions:
1. HotSpot: VM to native (`_thread_in_vm` -> `_thread_in_native`)
2. SVM: Native to VM (enter SVM)
3. SVM: VM to native (exit SVM)
4. HotSpot: Native to VM (`_thread_in_native` -> `_thread_in_vm`)

When processing a `HotSpotCompiledCode` object during code installation (i.e. `CodeInstaller::install`), the overhead of all these transitions is significant.
This PR changes the way code installation is done by serializing a `HotSpotCompiledCode` object to a byte array (in malloc'ed memory). The C++ code then deserializes this format to install the final code in the code cache. The bulk of this change is in `jdk.vm.ci.hotspot.HotSpotCompiledCodeStream` and `jvmciCodeInstaller.cpp`.

There are other smaller changes to reduce JNI overhead included in this PR:
* Pass raw VM values alongside JVMCI wrapper objects to the VM to avoid JNI upcall to unbox in C++. For example, the [getBytecode](https://github.com/openjdk/jdk/blob/33369719b2e39bddd4a1b7f300f36506306b03fa/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/CompilerToVM.java#L125) native method passed a `HotSpotResolvedJavaMethodImpl` to the VM and the VM then subsequently needs a JNI upcall to extract the `Method*` value from the `HotSpotResolvedJavaMethodImpl.methodHandle` field. With PR, `getBytecode` now passes an argument pair (i.e. `getBytecode(HotSpotResolvedJavaMethodImpl method, long methodPointer);`) that removes the unboxing upcall. This is done for all JVMCI objects that wrap a C++ object pointer. The box object is still passed as it protects the validity of the C++ pointer value.
* Cache HotSpotConstantPool holder in [`HotSpotConstantPool.getHolder()`](https://github.com/openjdk/jdk/blob/33369719b2e39bddd4a1b7f300f36506306b03fa/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotConstantPool.java#L236).
* Remove VM round trip for converting a `Klass*` value to a `ResolvedJavaType`.
* Avoid eager `String` creation in `JVMCIEnv::get_jvmci_type`.

This change is primarily in JVMCI-only code. It has been extensively tested in GraalVM on JDK 17. On the [octane](https://developers.google.com/octane) JavaScript benchmark, this PR reduces total time spent by GraalVM EE in JVMCI code installation from 648 s to 382 s.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8289094](https://bugs.openjdk.org/browse/JDK-8289094): [JVMCI] reduce JNI overhead and other VM rounds trips in JVMCI

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9305/head:pull/9305` \
`$ git checkout pull/9305`

Update a local copy of the PR: \
`$ git checkout pull/9305` \
`$ git pull https://git.openjdk.org/jdk pull/9305/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9305`

View PR using the GUI difftool: \
`$ git pr show -t 9305`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9305.diff">https://git.openjdk.org/jdk/pull/9305.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/utilities/ostream.hpp [hunk 1]
```diff
@@ -100,7 +100,7 @@ class outputStream : public ResourceObj {
    void print_raw(const char* str, size_t len)   { write(str,         len); }
    void print_raw_cr(const char* str)         { write(str, strlen(str)); cr(); }
    void print_raw_cr(const char* str, size_t len){ write(str,         len); cr(); }
-   void print_data(void* data, size_t len, bool with_ascii);
+   void print_data(void* data, size_t len, bool with_ascii, bool rel_addr=true);
```
Comment (dougxc) [inline src/hotspot/share/utilities/ostream.hpp:103]: This allows for hex dumps to show absolute addresses which is important for comparing a buffer [written in Java](https://github.com/openjdk/jdk/pull/9305/files#diff-d290f5a7a572ba7d4f0c7d234eb73de6537a2aa1d02726527d8db0df23c874b7R118-R134) and [read in C++](https://github.com/openjdk/jdk/pull/9305/files#diff-5e4a9ef468fc0f1564d6cfbe98f44b1099842098366cbec4103710137b2022edR34-R36).

**Review by dean-long — APPROVED** [review]
I took a quick look and didn't find any problems.  It's clever how the tag values are synchronized.

**Review by dholmes-ora — COMMENTED** [review]
Hi Doug,

I just looked at the general runtime changes and they seem okay.

Thanks.

**Comment by dougxc** [comment]
Thanks @vnkozlov @dean-long and @dholmes-ora for the review. I'm doing a little more GraalVM testing before merging.

**Pull request closed without merging**